src/3d/qgs3dmodelcache.h

```diff
@@ -0,0 +1,119 @@
+/***************************************************************************
+                         qgs3dmodelcache.h
+                         ---------------
+    begin                : July 2020
+    copyright            : (C) 2020 by Mathieu Pellerin
+    email                : nirvn dot asia at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGS3DMODELCACHE_H
+#define QGS3DMODELCACHE_H
+
+#include "qgsabstractcontentcache.h"
+#include "qgis_sip.h"
+#include "qgis_core.h"
+#include "qgis_3d.h"
+
+#include <Qt3DRender/QSceneLoader>
+#include <Qt3DRender/QMesh>
+
+#ifndef SIP_RUN
+
+///@cond PRIVATE
+
+/**
+ * \ingroup 3d
+ * \class Qgs3DModelCacheEntry
+ * An entry for a Qgs3DModelCache, representing a 3D model encoded as Base64
+ * \since QGIS 3.6
+ */
+class _3D_EXPORT Qgs3DModelCacheEntry : public QgsAbstractContentCacheEntry
+{
+  public:
+
+    /**
+     * Constructor for Qgs3DModelCacheEntry, corresponding to the specified \a path.
+     */
+    Qgs3DModelCacheEntry( const QString &path ) ;
+
+    //! The 3D model data encoded as Base64
+    QByteArray base64;
+
+    int dataSize() const override;
+    void dump() const override;
+    bool isEqual( const QgsAbstractContentCacheEntry *other ) const override;
+
+};
+
+///@endcond
+#endif
+
+/**
+ * \class Qgs3DModelCache
+ * \ingroup 3d
+ * A cache for 3D models.
+ *
+ * Qgs3DModelCache delivers pre-fetched 3D models.
+ *
+ * Qgs3DModelCache is not usually directly created, but rather accessed through
+ * QgsApplication::3dModelCache().
+ *
+ * \since QGIS 3.16
+*/
+#ifdef SIP_RUN
+class _3D_EXPORT Qgs3DModelCache : public QgsAbstractContentCacheBase // for sip we skip to the base class and avoid the template difficulty
+{
+#else
+class _3D_EXPORT Qgs3DModelCache : public QgsAbstractContentCache< Qgs3DModelCacheEntry >
+{
+#endif
+    Q_OBJECT
+
+  public:
+
+    /**
+     * Constructor for Qgs3DModelCache, with the specified \a parent object.
+     */
+    Qgs3DModelCache( QObject *parent SIP_TRANSFERTHIS = nullptr );
+
+    /**
+     * Returns a QMesh object with its source set to the specified \a path
+     *
+     * \a path may be a local file, remote (HTTP) url, or a base 64 encoded string (with a "base64:" prefix).
+     */
+    Qt3DRender::QMesh *pathAsMesh( const QString &path, bool blocking = false );
```

Comment:
I'd suggest adding SIP_FACTORY to this and pathAsSceneLoader -- even though it's currently not exposed to sip, it may save hassle in the long term if this class is ever exposed. Either that or make it return a std::unique_ptr so that it's clear that ownership is returned